lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1295,25 +1327,111 @@ class MongoClientInterface {
 
+    /**
+     * Flags the object before deleting it, this is done
+     * to keep object metadata in the oplog, as oplog delete
+     * events don't contain any object metadata
+     * @param {Object} collection MongoDB collection
+     * @param {string} bucketName bucket name
+     * @param {string} key Key of the object to delete
+     * @param {object} filter additional query filters
+     * @param {Logger}log logger instance
+     * @param {Function} cb callback containing error
+     * and BulkWriteResult
+     * @return {undefined}
+     */
+    internalDeleteObject(collection, bucketName, key, filter, log, cb) {
+        // filter used when finding and updating object
+        const findFilter = Object.assign({
+            _id: key,
+            $or: [
+                { 'value.deleted': { $exists: false } },
+                { 'value.deleted': { $eq: false } },
+            ],
+        }, filter);
+        // filter used when deleting object
+        const updateDeleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, filter);
+        async.waterfall([
+            // Adding delete flag when getting the object
+            // to avoid having race conditions.
+            next => collection.findOneAndUpdate(findFilter, {
+                $set: {
+                    '_id': key,
+                    'value.deleted': true,
+                },
+            }, {
+                upsert: false,
+            }, (err, doc) => {
+                if (err) {
+                    log.error('internalDeleteObject: error getting object',
+                        { bucket: bucketName, object: key, error: err.message });
+                    return next(errors.InternalError);
+                }
+                if (!doc.value) {
+                    log.error('internalDeleteObject: unable to find target object to delete',
+                        { bucket: bucketName, object: key });
+                    return next(errors.NoSuchKey);
+                }
+                const obj = doc.value;
+                const objMetadata = new ObjectMD(obj.value);
+                objMetadata.setOriginOp('s3:ObjectRemoved:Delete');
+                objMetadata.setDeleted(true);
+                return next(null, objMetadata.getValue());
+            }),
+            // We update the full object to get the whole object metadata
+            // in the oplog update event
+            (objMetadata, next) => collection.bulkWrite([
+                {
+                    updateOne: {
+                        filter: updateDeleteFilter,
+                        update: {
+                            $set: { _id: key, value: objMetadata },
+                        },
+                        upsert: false,
+                    },
+                }, {
+                    deleteOne: {
+                        filter: updateDeleteFilter,
+                    },
+                },
+            ], { ordered: true }, next),
```

Comment:
We might not want to return an error if the `collection.bulkWrite()` fails.
At this stage, the operation is already considered "completed" with `deleted:true` set.
Also, the user/client won't be able to retry the operation in case of failure at this stage and it would assume the operation has failed.

Reply:
should be good now